File: base/src/main/java/com/boilerplate/base/mvp/BaseActivity.java
package com.boilerplate.base.mvp;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseActivity<T extends IPresenter> extends AppCompatActivity implements IView {

    protected T presenter;
    protected ProgressDialog progressDialog;
    private int resId;
    protected List<Fragment> fragmentBackStack;
    private boolean isAnimationRunning = false;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayout());
        fragmentBackStack = new ArrayList<>();
        presenter = getPresenter();
        setupActivityComponent();
        if (presenter != null) presenter.attachView(this);
        initEventAndData();
    }

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onStop() {
        super.onStop();
    }


    protected abstract T getPresenter();


    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (presenter != null) presenter.detachView();
    }

    protected abstract int getLayout();

    protected abstract void initEventAndData();

    protected abstract void setupActivityComponent();
}

File: base/src/main/java/com/boilerplate/base/kotlin/BaseActivity.kt
package com.boilerplate.base.kotlin

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

/**
 * Created by Louis on 2018/4/27.
 */
abstract class BaseActivity<T, S> : AppCompatActivity() where T : IPresenter<S>, S : IView {

    protected abstract var mPresenter: T
    protected abstract var mView: S
    protected abstract val layout: Int


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(layout)
        setupActivityComponent()
        mPresenter.attachView(mView)
        initEventAndData()
    }

    override fun onDestroy() {
        super.onDestroy()
        mPresenter.detachView()
    }

    protected abstract fun getLayout()

    protected abstract fun setupActivityComponent()

    protected abstract fun initEventAndData()
}